docs/development/design/embed-api.rst

```diff
@@ -0,0 +1,211 @@
+It has some known problems:
+
+* There are different ways of querying the API: ``?url=`` (generic) and ``?doc=`` (relies on Sphinx's specific concept)
+* Doesn't support MkDocs
+* Lookups are slow (~500 ms)
+* IDs returned aren't well formed (like empty IDs ``"headers": [{"title": "#"}]``)
+* The content is always an array of one element
+* It tries different variations of the original ID
+* It doesn't return valid HTML for definition lists (``dd`` tags without a ``dt`` tag)
+
+
+Goals
+-----
+
+Considering the problems mentioned in the previous section,
+the inclusion of new features and the definition of a contract that works the same for all,
+this document set the following goals for the new version of this endpoint:
+
+* Support external documents hosted outside Read the Docs
+* Do not depend on Sphinx ``.fjson`` files
+* Query and parse the ``.html`` file directly (from our storage or from an external request)
+* Rewrite all links returned in the content to make them absolute
+* Always return valid HTML structure
+* Delete HTML tags from the original document if needed
+* Support ``?nwords=`` and ``?nparagraphs=`` to return chunked content
+* Require a valid HTML ``id`` selector
+* Handle special cases for particular doctools (e.g. Sphinx requires to return the ``.parent()`` element for ``dl``)
+* Make explicit the client is asking to handle the special cases (e.g. send ``?doctool=sphinx&version=4.0.1``)
+* Accept only ``?url=`` request GET argument to query the endpoint
+* Add HTTP cache headers to cache responses
+* Allow :abbr:`CORS` from everywhere
+
+
+Embed endpoint
+--------------
+
+Returns the exact HTML content for a specific identifier.
+If no anchor identifier is specified the content of the whole page is returned.
+
+.. http:get:: /api/v3/embed/?url=https://docs.readthedocs.io/en/latest/development/install.html#set-up-your-environment
+
+   :query url (required): Full URL for the documentation page with optional anchor identifier.
+   :query expand (optional): Allows to return extra data about the page. Currently, only ``?expand=identifiers`` is supported
+      to return all the identifiers that page accepts.
+
+   .. sourcecode:: json
+
+      {
+         "project": "docs",
+         "version": "latest",
+         "language": "en",
+         "path": "development/install.html",
+         "title": "Development Installation",
+         "url": "https://docs.readthedocs.io/en/latest/install.html#set-up-your-environment",
+         "id": "#set-up-your-environment",
+         "content": "<div class=\"section\" id=\"development-installation\">\n<h1>Development Installation<a class=\"headerlink\" href=\"https://docs.readthedocs.io/en/stable/development/install.html#development-installation#development-installation\" title=\"Permalink to this headline\">¶</a></h1>\n ..."
+      }
+
+
+   When used together with ``?expand=identifiers`` the follwing field is also returned:
```

Comment:
I'm not sure about returning this information on the same endpoint, is confusing returning the content of a section plus the list of sections, we should have a different endpoint to return data about the page itself.

Reply:
Can you expand here why it may cause confusion and how having a different endpoint to only show the available `id=` for a page is better? I think I'm not against a different endpoint, but I want to understand your position better.

Reply:
Because you are requesting information for a section, but getting information about the page too, you don't want to have to query a random section just to get information about the page. This also allow us to return more information about the page itself like the title.

Reply:
Pinging @agjohnson here that he wrote:

> More metadata around headers, such as heading level. I'd like to display the topics in a nested menu, as they don't make as much sense in sequential order

in https://github.com/readthedocs/readthedocs.org/issues/7117

@agjohnson can you take a look at this proposal and suggest that would be the ideal response to you and how you would like to use it?

Reply:
@agjohnson returning headers may not be useful if those headers don't have an `id=` that we can find. Also, I realized that it's not trivial to "find the title for a particular `id=`" ([we have this hack currently implemented that only works over `a` tags](https://github.com/readthedocs/readthedocs.org/blob/0340604163a15fccc8b786ce0a5e589bf78f6f24/readthedocs/embed/views.py#L283-L286). The `id=` could be in the `div` surrounding the `h1` tag or the `dl` tag or any other.

```html
<div id="configuration">
  <h1>Configuration<a class="headerlink" href="#configuration" title="Permalink to this headline">¶</a></h1>
  <p>...</p>
</div>
```

For this case, we could get the `.next()` from the `id=configuration`. However, we can't guarantee that this will always work. Even if we get it, we will want to remove the trailing `¶` from its name as well in this case (but it could be a different char)


```html
<dt id="confval-hoverxref_modal_class">
  <code class="sig-name descname">hoverxref_modal_class</code>
  <a class="headerlink" href="#confval-hoverxref_modal_class" title="Permalink to this definition">¶</a>
</dt>
```

This one is similar, but the `.next()` element is exactly the title we are looking for. It does not contain the `¶` char.

However, if the title we want is exactly on the `h1` or `dl` (instead of their child), we will fail to detect it and we would return something invalid.

----

So, I think it's still useful to return all the available `id=`s from the page so developers can explore them, but it's not easy (maybe impossible) to know the exact title for that particular `id=` and the exact hierarchy as well.

Reply:
@agjohnson I was asking about how is the ideal output you would like to get from the API. Besides, I raised some issues that will make it hard to do it _right_ with the output that I imagined you need. Can you take a look at my previous comments and tell me your thoughts on this?

Reply:
I moved this to a particular endpoint `/api/v3/embed/identifiers/` and define the initial response: "return all the available identifiers for a specific page".

We can expand it later if we need something else and there is a good way to do it.

Reply:
We talked more here, and the use case or feature that I'm describing is currently in a strange place.

More explicitly, I require a list of headings -- basically the toctree on a document -- with heading text, URL to link to the heading, and the heading nesting level or heading nesting as a data structure. This gives me what I need to basically expose the toctree in our application. This is the feature that I had wanted to expose for commercial use -- embedding documentation metadata/headings/etc in customer applications. A hoverxref type extension might be useful in addition here, but is separate. Customers would still need to get metadata out of a particular document in order to inject a toctree into an application view.

So, where I'm at is that I feel like this feature is more of a relic of where the embed API started and it is dragging the direction of the embed API down -- however this might just be my interpretation. The embed API is more focused on having generic support, and so therefore parsing HTML, and what I want is basically exposing the contents of `objects.inv`. I could be talking about a separate feature, or we could still be talking about keeping this as a separate API endpoint.

I am going to research this a bit more, I may be talking about a separate feature entirely at this point.

Reply:
Yup, so we talked about exposing a separate API for that instead.

We need to decide if this feature is going to be a separate product and how users will interact with it. I think we'd be maintaining 1 client JS library that does both embed of hover cards and injection of document header/refs from Sphinx, as the use case is similar.

@humitos Did we talk about what mechanism we'd use if this function was a subpath on the embed API? Are we using Sphinx refs via the SphinxDomain modeling?

Reply:
> @humitos Did we talk about what mechanism we'd use if this function was a subpath on the embed API? Are we using Sphinx refs via the SphinxDomain modeling?

Nope, we didn't talk about this and I haven't thought too much either. I usually forget that we have `SphinxDomain` model --and I haven't used it yet.

Something like `/api/v3/embed/headers/` may work for this use case. However, we have to keep in mind that:

- it will only work with Sphinx if we use the `SphinxDomain` model
- making it generic by parsing the HTML is probably impossible
- we can use the HTTP request arguments for other doctools, as `?doctool=mkdocs&version=1.0.1` and parse a known HTML structure